dev/core/#1532 -  Adds warning similar to 5.57 if Activity Revisions are present.

stesi561 opened the pull request:
This might be either because they have been added since then or that they were never addressed by the user.

Overview
----------------------------------------
Activity Revisions were removed in 5.57 however there were a couple of lingering references that caused additional revisions to be created. These have been addressed and now we need to rerun the upgrade to detect and warn the user about them.

seamuslee001 commented on CRM/Upgrade/Incremental/php/FiveSixtyTwo.php:
```diff
@@ -39,6 +39,11 @@ public function setPreUpgradeMessage(&$preUpgradeMessage, $rev, $currentVer = NU
+
+    $activityCount = CRM_Core_DAO::singleValueQuery('SELECT MAX(id) FROM civicrm_activity');
+    if ($activityCount > 0) {
+      $preUpgradeMessage .= '<p>' . ts('Your database contains CiviCase activity revisions which are deprecated and will begin to appear as duplicates in SearchKit/api4/etc.<ul><li>For further instructions see this <a %1>Lab Snippet</a>.</li></ul>', [1 => $docAnchor]) . ts('Note: You might have followed these steps already but unfortunately a previous upgrade which removed CiviCase Activity revis ...
```
@stesi561 $docAnchor isn't defined here you may need to bring that in

demeritcowboy replied:
I don't want to get bogged down in wording but the upgrade in 5.54 didn't remove revisions, it "attempted to no longer created new revisions". So maybe "...a previous upgrade which started ignoring the setting to create CiviCase Activity revisions failed to account..."

demeritcowboy commented on CRM/Upgrade/Incremental/php/FiveSixtyTwo.php:
```diff
@@ -39,6 +39,13 @@ public function setPreUpgradeMessage(&$preUpgradeMessage, $rev, $currentVer = NU
       }
     }
+
+    $activityCount = CRM_Core_DAO::singleValueQuery('SELECT MAX(id) FROM civicrm_activity');
```
Thanks for doing this. We only want to display this rerun though if they actually have old revisions in the db. So suggest:
```suggestion
    $activityCount = CRM_Core_DAO::singleValueQuery('SELECT MAX(id) FROM civicrm_activity WHERE is_current_revision = 0');
```

stesi561 commented:
Sorry about the delay I was doing some thinking about this and wonder if I'm now overthinking this.

I think there are the following matrix

0) No new activity revisions have been created.
1) New activity revisions have been created.

A) Index was removed automatically
B) Index was removed manually
C) Index wasn't removed manually

ξ) Activity revisions were tidied up
ζ) Activity revisions were not tidied up

I think we just need to think about ξ, ζ in terms of language but not account for a different process.

I guess a question is:
a) Is it okay to just redo the complete logic of the previous upgrade which was
   i) Detect if too many activities to update the index as part of the upgrade
  ii) If below limit then update the index
 iii) warn user about need to handle the activity revisions, and/or update the index.

Or should we try and detect if the index was still in place. I dived a bit down the rabbit hole on how hard this was going to be.

In CRM/Core/BAO/SchemaHandler.php there are several likely looking functions. We should be able to use getIndexName to get the index name then something like: 

<pre>
select * from information_schema.REFERENTIAL_CONSTRAINTS as rc where CONSTRAINT_SCHEMA = DATABASE() and CONSTRAINT_NAME = %1 and DELETE_RULE = 'SET NULL';
</pre>
which if returns more than 0 rows we should alert the user. but is this overkill?

After getting to this point I then consider - do we just want to keep it simple and leave it to the user to resolve?

stesi561 committed:
Adds warning similar to 5.57 if Activity Revisions are present.

Checks if upgrade is from version after 5.57.alpha1

If so check for any activities which are not the current revision

Issue warning.

Notes:
- Doesn't check if Foreign Key on delete trigger was updated to prevent
  deletion
- Doesn't warn about the above - we assume this is covered in the steps
  people will follow in the link - or that they will have done this
  already.

demeritcowboy commented on CRM/Upgrade/Incremental/php/FiveSixtyTwo.php:
```diff
@@ -23,6 +23,19 @@ class CRM_Upgrade_Incremental_php_FiveSixtyTwo extends CRM_Upgrade_Incremental_B
 
   public function setPreUpgradeMessage(&$preUpgradeMessage, $rev, $currentVer = NULL) {
     if ($rev == '5.62.alpha1') {
+
```
Of course when you originally wrote this it was alpha1, now it's beta1.
```suggestion
if ($rev == '5.62.beta1') {
```

stesi561 replied:
Confused? I still see 5.62.alpha1 here -> https://github.com/civicrm/civicrm-core/blob/625954e854d6ec2c34f6c40e52d71cf69670f7b5/CRM/Upgrade/Incremental/php/FiveSixtyTwo.php#LL25C18-L25C29
and 
https://github.com/civicrm/civicrm-core/blob/master/CRM/Upgrade/Incremental/php/FiveSixtyTwo.php#LL25C18-L25C29

Or do I need to account for the fact that we'll be making a new revisions here? In which case do I just add my upgrade logic in a different conditional to the existing?

demeritcowboy replied:
Oh right, sorry I didn't see there was already other stuff in the function. You'll need to add it to move it to an elseif clause, `elseif ($rev == '5.62.beta1') {`

demeritcowboy commented on CRM/Upgrade/Incremental/php/FiveSixtyTwo.php:
```diff
@@ -23,6 +23,19 @@ class CRM_Upgrade_Incremental_php_FiveSixtyTwo extends CRM_Upgrade_Incremental_B
+          // Text copied from FiveFifty Seven
+          $preUpgradeMessage .= '<p>' . ts('Your database contains CiviCase activity revisions which are deprecated and will begin to appear as duplicates in SearchKit/api4/etc.<ul><li>For further instructions see this <a %1>Lab Snippet</a>.</li></ul>', [1 => $docAnchor]) . '</p>';
+          // New text explaination as to why we show this again.
+          $preUpgradeMessage .= '<p>' . ts('Note: You might have followed these steps already but unfortunately a previous upgrade which removed CiviCase Activity revisions failed to account for all the places this functionality existed. This means that either new Revisions have been added since you removed them, or not all were removed.') . '</p>';
```
I still think the word "removed" might cause some confusion - nothing was removed by the upgrade. If you want to keep the word "removed" then how about "...a previous upgrade which told you to remove CiviCase Activity revisions..."?

stesi561 replied:
sorry yes totally will address.

stesi561 commented:
Fixed up the above and rebased into a single commit.

demeritcowboy commented:
Thanks @stesi561 and for sticking with it. Will give it a little r-run tomorrow.
Like the idea to restrict to 5.57+ upgrades.

demeritcowboy commented:
Ah I forgot that since there's no actual tasks for 5.62.beta1 it doesn't call the preupgrade for it. It also needs something like this, but otherwise looks good and is mergeable after adding:

```php
  /**
   * This needs to exist otherwise it doesn't run the preupgrade message.
   */
  public function upgrade_5_62_beta1($rev): void {
  }
```

stesi561 committed:
Add empty upgrade step to force preUpgrade to run

demeritcowboy commented:
Thanks @stesi561 for the final "revision" to this PR.
